octoprint_mrbeam/iobeam/laserhead_handler.py

```diff
@@ -0,0 +1,140 @@
+
+import os
+import yaml
+from octoprint_mrbeam.mrb_logger import mrb_logger
+
+# singleton
+_instance = None
+
+
+def laserheadHandler(plugin):
+	global _instance
+	if _instance is None:
+		_instance = LaserheadHandler(plugin)
+	return _instance
+
+
+class LaserheadHandler(object):
+	LASER_POWER_GOAL = 950
+
+	def __init__(self, plugin):
+		self._logger = mrb_logger("octoprint.plugins.mrbeam.iobeam.laserhead")
+		self._plugin = plugin
+		self._settings = plugin._settings
+		self._plugin_version = plugin._plugin_version
+
+		self._lh_cache = {}
+		self._last_used_lh_serial = None
+		self._correction_settings = {}
+		self._laser_heads_file = os.path.join(self._settings.getBaseFolder("base"), self._settings.get(['laser_heads', 'filename']))
+		self._load_laser_heads_file()  # Loads correction_settings, last_used_lh_serial and lh_cache
+
+		self._current_used_lh_serial = self._last_used_lh_serial
+
+	def set_current_used_lh_serial(self, serial):
+		if serial:
+			self._current_used_lh_serial = serial
+			self._write_laser_heads_file()
+
+			if serial not in self._lh_cache:
+				self._logger.info('Saving new laser head serial number: {}'.format(serial))
+				self._lh_cache[serial] = dict(correction_factor=1)
+
+	def set_power_measurement_value(self, measure, value):
+		if measure not in self._lh_cache[self._current_used_lh_serial]:
```

Comment:
please ensure that you have a serial num or if not, throw an exception / write error to the log

Reply:
done